src/stash/mod.rs

```diff
@@ -364,20 +364,41 @@ impl<V> Stash<V> {
 
     /// Take an item from a slot (if non empty).
     pub fn take(&mut self, index: usize) -> Option<V> {
-        if let Some(entry) = self.data.get_mut(index) {
-            match mem::replace(entry, Entry::Empty(self.next_free)) {
+        match self.data.get_mut(index) {
+            None => None,
+            Some(entry) => match mem::replace(entry, Entry::Empty(self.next_free)) {
                 Entry::Empty(free_slot) => {
-                    // Just put it back.
                     *entry = Entry::Empty(free_slot);
-                }
+                    None
+                },
                 Entry::Full(value) => {
                     self.next_free = index;
                     self.size -= 1;
-                    return Some(value);
+                    Some(value)
                 }
             }
         }
-        None
+    }
+
+    /// Take an item from a slot (if non empty) without bounds checking.
+    /// So use it very carefully!
+    /// 
+    /// This can be safely used as long as the user does not mutate
+    /// `indices` from `put` and is sure not to have taken the value
+    /// associated with the given `index`.
+    pub unsafe fn take_unchecked(&mut self, index: usize) -> Option<V> {
+        let entry = self.data.get_unchecked_mut(index);
+        match mem::replace(entry, Entry::Empty(self.next_free)) {
+            Entry::Empty(free_slot) => {
```

Comment:
In the truly unchecked case, you'd mark this case as `unreachable()`. Ditto for the other unchecked functions.